tests/end_to_end/test_control_flow.py

```diff
@@ -38,3 +41,292 @@ def test_while_executes_and_scopes_correctly(execute):
     res = execute(SAMPLE_WHILE, artifacts=["result"])
     assert res.values["result"] == 6
     assert res.artifacts["result"] == SAMPLE_WHILE
+
+
+def test_if_should_slice_within_if(execute):
+    # TODO: We only slice within a block if we can execute it. Later, we need to be able to ensure we can perform slicing among unvisited blocks as well
+    CODE = """a = 10
+b = 20
+if a > 5:
+    a = 5
+    b = 6
+"""
+    res = execute(CODE, artifacts=["a", "b"])
+    assert res.values["a"] == 5
+    assert res.values["b"] == 6
+    assert (
+        res.artifacts["a"]
+        == """a = 10
+if a > 5:
+    a = 5
+"""
+    )
+
+
+@pytest.mark.xfail(
+    reason="Variable overwritten in visited branch results in declaration outside slice being removed"
+)
+def test_if_should_not_slice_out_existing_variable_definitions_outside_block(
+    execute,
+):
+    CODE = """a = 10
+b = 20
+if a > 5:
+    a = 5
+    b = 6
+"""
+    res = execute(CODE, artifacts=["a", "b"])
+    assert res.values["a"] == 5
+    assert res.values["b"] == 6
+    assert (
+        res.artifacts["b"]
+        == """a = 10
+b = 20
+if a > 5:
+    b = 6
+"""
+    )
+
+
+def test_if_should_slice_within_else(execute):
+    # TODO: We only slice within a block if we can execute it. Later, we need to be able to ensure we can perform slicing among unvisited blocks as well
+    CODE = """a = 10
+b = 20
+if a <= 5:
+    a = 5
+    b = 5
+else:
+    a = 100
+    b = 101
+"""
+    res = execute(CODE, artifacts=["a", "b"])
+    assert res.values["a"] == 100
+    assert res.values["b"] == 101
+    assert (
+        res.artifacts["a"]
+        == """a = 10
+if a <= 5:
+    a = 5
+    b = 5
+else:
+    a = 100
+"""
+    )
+    assert (
+        res.artifacts["b"]
+        == """a = 10
+if a <= 5:
+    a = 5
+    b = 5
+else:
+    b = 101
+"""
+    )
+
+
+def test_if_should_not_slice_whole_block_out_if_unexecuted_block_present(
+    execute,
+):
+    CODE = """a = 10
+b = 20
+if a >= 10:
+    a += 1
+else:
+    a -= 1
+"""
+    res = execute(CODE, artifacts=["b"])
+    assert (
+        res.artifacts["b"]
+        == """a = 10
+b = 20
```

Comment:
The reason we're not slicing off the whole `if` block is that the unexec block may contain something related to `b`, which we do not know since the branch never got executed. (Note in the earlier behavior too, we would not know what is happening in any code which was not executed).
In case there is no unexec block, the whole `if` block will be sliced out, there's a test for that (`test_if_slice_out_whole_block_if_no_unexecuted_branch`).
This test will be changed when static analysis is included.

Reply:
By this logic, we should keep everything since there is no telling what could have happened in any blackbox, which is a fair assumption for blackboxes. I think we used to have a solution that at least accounts for variable names in blackboxes to understand the scope of things that could be impacted.

Reply:
The prior solutions only worked if the instructions were visited. If there was an instruction in a blackbox that was never visited, like in the if/else scenario, we were not able to do any analysis, and a similar behavior would occur. In such a cse, I am trying to ensure we support everything we did in the past and building up on atleast some, if not all of the possible scenarios we could encounter.